[esy] Read dependencies in esy from installation.json

ulrikstrid opened the pull request:
I'm piggy backing on the `custom_resolution` flag that is already set when building with esy here. If we can't find the dependency in the environment variables we'll look it up in the `installation.json` file from the current sandbox.

@bobzhang can I make the lookup per package be lazy somehow so that we don't have to do it multiple times?

Notes:

- This implementation assumes that the build is run with `"buildsInSource": true` which I think is valid and needed anyway.
- This PR doesn't make much sense without landing some variation of #3856

bsansouci commented on jscomp/bsb/bsb_pkg.ml:
```diff
@@ -106,6 +106,22 @@ let pkg_name_as_variable package =
   |> fun s -> Ext_string.split s '-'
   |> String.concat "_"
 
+let rec find_dep_path ic package_name =
```
One thing you could do that might make this function simpler is make this function do the try/catch and return an option of string

ulrikstrid replied:
Good idea!

bobzhang commented:
> @bobzhang can I make the lookup per package be lazy somehow so that we don't have to do it multiple times?

Yeah, I think it is a good idea

The PR is small and looks nonintrusive, lgtm

ulrikstrid commented:
@bobzhang I couldn't figure out how to make the computation lazy but I moved the try/catch to a better position and I think this is good to go now.

I also rebased on master.
